src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,1214 @@
+                            // avail ring, for later processing.
+                            queue.undo_pop();
+                            break;
+                        }
+                    }
+                    let status = match request.execute(
+                        &mut self.disk_image,
+                        self.disk_nsectors,
+                        &self.mem,
+                        &self.disk_image_id,
+                    ) {
+                        Ok(l) => {
+                            len = l;
+                            VIRTIO_BLK_S_OK
+                        }
+                        Err(e) => {
+                            error!("Failed to execute request: {:?}", e);
+                            METRICS.block.invalid_reqs_count.inc();
+                            len = 1; // We need at least 1 byte for the status.
+                            e.status()
+                        }
+                    };
+                    // We use unwrap because the request parsing process already checked that the
+                    // status_addr was valid.
+                    self.mem
+                        .write_obj_at_addr(status, request.status_addr)
+                        .unwrap();
+                }
+                Err(e) => {
+                    error!("Failed to parse available descriptor chain: {:?}", e);
+                    METRICS.block.execute_fails.inc();
+                    len = 0;
+                }
+            }
+            queue.add_used(&self.mem, head.index, len);
+            used_any = true;
+        }
+
+        used_any
+    }
+
+    pub(crate) fn signal_used_queue(&self) -> result::Result<(), DeviceError> {
+        self.interrupt_status
+            .fetch_or(VIRTIO_MMIO_INT_VRING as usize, Ordering::SeqCst);
+
+        self.interrupt_evt.write(1).map_err(|e| {
+            error!("Failed to signal used queue: {:?}", e);
+            METRICS.block.event_fails.inc();
+            DeviceError::FailedSignalingUsedQueue(e)
+        })?;
+        Ok(())
+    }
+
+    /// Update the backing file for the Block device
+    pub fn update_disk_image(&mut self, disk_image: File) -> result::Result<(), DeviceError> {
+        self.disk_image = disk_image;
+        self.disk_nsectors = self
+            .disk_image
+            .seek(SeekFrom::End(0))
+            .map_err(DeviceError::IoError)?
+            / SECTOR_SIZE;
+        self.disk_image_id = build_disk_image_id(&self.disk_image);
+        METRICS.block.update_count.inc();
+        Ok(())
+    }
+}
+
+impl VirtioDevice for Block {
+    fn device_type(&self) -> u32 {
+        TYPE_BLOCK
+    }
+
+    fn get_queues(&mut self) -> &mut Vec<Queue> {
+        &mut self.queues
+    }
+
+    fn get_queue_events(&self) -> Vec<EventFd> {
+        vec![self
+            .queue_evt
+            .try_clone()
+            .expect("Unable to clone queue event fd.")]
```

Comment:
Ideally, we shouldn't `expect()` here. Users of this crate might want to be able to recover from this error.

Reply:
The error needs to be very specific and actionable. It is very hard to make it so in this case, and even then If that call fails it will render the block device partially configured or otherwise in an undefined state so the user might not be able to do anything to work around without additional support.

Reply:
Ideally it would be specific and actionable, but the `EventFd` error we have everywhere else in our code will also suffice.

The big drawback of `expect()` is that it will kill the whole process. To end the whole process because of an error a very big assumption to make regarding the importance of that error in the process.

I believe the best practice is for all components (especially leaf components) to propagate errors up the tree, with `expect()`s ideally only in the root. Of course there's exceptions here like poisoned locks, unexpectedly closed channels, etc, where one can safely assume (no matter the level of the tree) that the integrity of the process is compromised.

Reply:
Fixed.